### PR TITLE
core : support output vocab size distinct from embedding vocab size

### LINKED ISSUE #26224: core: support output vocab size distinct from embedding vocab size
## Problem

`llama_vocab::n_tokens()` is currently the single source of truth for both:
- the embedding table row count (`tok_embd`)
- the logits/sampling buffer width, via the public API contract documented at
  `llama_get_logits(ctx) + output_ids[i]*n_vocab` (include/llama.h)

This assumes embedding rows == output rows, which holds for every architecture
currently in the tree. It does not hold for swiss-ai/Apertus-v1.5-8B: its
`embed_tokens.weight` is `[266752, 4096]` while `lm_head.weight` is a separate,
untied `[131072, 4096]` tensor. The extra 135680 ids in the embedding table are
reserved for multimodal (image/audio) input-only tokens that are never valid
model outputs - confirmed against the model's own tokenizer_config.json
(`base_vocab_size: 131072`, matching lm_head exactly; the rest are "added
tokens").

Loading such a model today either fails outright (output tensor shape
mismatch) or requires fabricating ~135680 output rows that don't exist in the
checkpoint, just to keep n_vocab symmetric.

## Proposed fix

Mirror the existing `n_embd_out_impl` / `LLM_KV_EMBEDDING_LENGTH_OUT` pattern
(already used for output embedding dimension overrides):

- new optional GGUF key `%s.vocab_size_out`
- `hparams.n_vocab_out_impl` (default 0 = no override)
- `llama_model::n_vocab_out()` returns `n_vocab_out_impl > 0 ? n_vocab_out_impl
  : vocab.n_tokens()`
- the ~5 buffer-sizing call sites in `llama-context.cpp` (`decode()`,
  `encode()`, `output_reserve()`, `output_reorder()`, backend-sampler init)
  switch from `vocab.n_tokens()` to `model.n_vocab_out()`

For every existing model this is a no-op (n_vocab_out_impl stays 0), so
behavior is unchanged unless a converter explicitly opts in.

## Status

Implemented and tested locally: full CPU `ci/run.sh` (debug+release ctest,
model-based smoke tests) passes 100%, plus manual regression testing against
a real model (Qwen2.5-0.5B-Instruct) confirming no behavior change. PR to
follow.

This is a prerequisite for Apertus v1.5 support, split out as its own PR per
the "separate PRs per feature" guideline in CONTRIBUTING.md.

### PR DESCRIPTION
**Pull request by adithyab94**
## Overview

`llama_vocab::n_tokens()` controls both the embedding table and the logits/sampling buffers. They have identical size for all current arches, and were thus never separated, but the relationship is also baked in the public API (`include/llama.h` logits stride `output_ids[i]*n_vocab`).

[Apertus v1.5 (`swiss-ai/Apertus-v1.5-8B`)](https://huggingface.co/swiss-ai/Apertus-v1.5-8B) consumes text/image/audio tokens but outputs only text, meaning its `lm_head` is indeed smaller than the embedding table (confirmed from `config.json` of the checkpoint plus `.safetensors` header):

```
vocab_size = 266752   # embed_tokens -> [266752, 4096]
output_vocab_size = 131072   # lm_head      -> [131072, 4096]   (untied)
```

The 135,680 excess rows are control and discrete image/audio tokens which occur only in input ID space, never as output. Unadulterated llama.cpp will either reject the `lm_head` (mismatched shape), or (in case of forcing) allocate the logits buffer for the bigger embedding vocabulary, thus corrupting data on every subsequent `ggml_backend_tensor_get_async` fill operation.

**Fix:** Use the same `n_embd_out_impl` / `LLM_KV_EMBEDDING_LENGTH_OUT` strategy for vocab size. Adds an optional GGUF key `%s.vocab_size_out` and `llama_model::n_vocab_out()` (= `n_vocab_out_impl > 0 ? n_vocab_out_impl : vocab.n_tokens()`). Now the few logits/sampling buffer allocation points in `llama-context.cpp` use `n_vocab_out()` while `vocab.n_tokens()` keeps controlling the embedding

## Additional information

Apertus v1.5 is the only consumer today. I can't name a second in-tree model that needs it. Implemented as a minimal generic mechanism rather than an Apertus-specific hack because that was cheaper and safer in `llama-context.cpp`, not because it's built ahead of demand. Cost is effectively zero at runtime. the one real tradeoff is a layer of indirection (`n_vocab_out()` vs `vocab.n_tokens()`) in a hot shared file. happy to reshape or narrow if preferred.

## Requirements

- [x] I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- [x] AI usage disclosure: Yes, it was predominately AI written by Sonnet 5 with steering.

### CODE REVIEW

**Comment by adithyab94** [comment]
PR was drafted with AI and made some changes to it

File: src/llama-model.h [hunk 1]
```diff
@@ -654,6 +654,9 @@ struct llama_model {
     uint32_t n_gpu_layers() const;
     llama_split_mode split_mode() const;
 
+    // output/logits vocab size (see llama_hparams::n_vocab_out_impl)
+    uint32_t n_vocab_out() const;
```
Comment (CISC) [inline src/llama-model.h:658]: This should probably live in `hparams` instead.

File: src/llama-model.cpp [hunk 2]
```diff
@@ -1707,6 +1708,10 @@ uint32_t llama_model::n_gpu_layers() const {
     return params.n_gpu_layers >= 0 ? params.n_gpu_layers : hparams.n_layer_all + 1;
 }
 
+uint32_t llama_model::n_vocab_out() const {
+    return hparams.n_vocab_out_impl > 0 ? hparams.n_vocab_out_impl : vocab.n_tokens();
+}
```
Comment (CISC) [inline src/llama-model.cpp:1713]: Correct me if I'm wrong, but won't we need to access this in `mtmd`? In that case you have to implement this as `llama_model_n_vocab_out` anyway, so better to leave this in `hparams`.

**Review by CISC — COMMENTED** [review]
@ggerganov / @ngxson any thoughts?

**Comment by ngxson** [comment]
I think this can be useful for certain models, especially audio backbone models that has different vocab than input vocab.

For example, qwen3tts simply pads the output logits to match vocab size:

https://github.com/ggml-org/llama.cpp/blob/34e9ee57f59d680db718e3a3466caceef7c24932/src/models/qwen3vl.cpp#L174-L188

However, changes in the current PR seems to be incomplete, a lot of things missing:
- public llama.h API
- what if model has a completely different set of output tokens? for ex, the same ID 0 can be mapped to <bos> input, but can be mapped something else output?
- need updates to common/sampling?

While the idea is valid, unless all details and edge cases are listed out, I don't think we can accept the implementation as-is.

Also, just to note that, even without this API, you can technically still implement the mentioned model by moving the audio/image tokens to mtmd, and slice them off the input embd tensor. Example: https://github.com/ggml-org/llama.cpp/pull/26190

**Comment by MichelRosselli** [comment]
Hi, I've been working on a broader Apertus v1.5 integration here:

https://github.com/MichelRosselli/llama.cpp/tree/model/apertus-v1.5

There is some overlap with this PR around the separate output vocabulary support. My branch currently handles the Apertus reduced output vocabulary with model-specific `-FLT_MAX` logit padding; the rest of the work covers chat and multimodal support.

I wanted to coordinate before opening my draft PR. Please let me know how you would prefer to handle the overlapping part.

I'm also using AI assistance during development, which I'll disclose in the PR.

**Comment by adithyab94** [comment]
Hi @MichelRosselli 

Thanks for the comment. Unfortunately, I can't work on this for the next couple of days due to some deadlines. If it's urgent, would you like to fix the suggestions and resolve the issues the reviewers mentioned?

**Comment by MichelRosselli** [comment]
Thanks! I don't think there's any urgency on my side. I'd like to first better understand the intended direction before figuring out how I might contribute to #26226.

Qwen3TTS already handles a similar input/output vocab mismatch upstream with model-specific padding, and I'm currently doing the same for Apertus in my branch, although the valid output range is positioned differently.

@ngxson, do you think it's worth defining a generic representation for this in the llama.cpp core already, including the output-token mapping, or would you prefer to keep these model-specific for now until we have more use cases and clearer requirements?

For Apertus I can keep the current `-FLT_MAX` padding approach for now, and then decide how best to proceed based on that direction.

**Comment by RaphaelKreft** [comment]
Hello @MichelRosselli, @adithyab94 and maintainers,

glad to see such nice community contributions for Apertus 1.5 : )

**Just as a heads-up:** we are currently still in the process of upstreaming Apertus 1.5 to transformers, so the checkpoint layout and reference implementation are not final.

So I would suggest to wait with the final merge until Apertus 1.5 is upstreamed and thus the reference implementation and the checkpoint are fixed (to cross check, validate and potentially port improvements and bug-fixes).

The transformers upstream PR: https://github.com/huggingface/transformers/pull/47662

**Comment by MichelRosselli** [comment]
Thanks for the heads-up! That sounds good to me :)

I'll keep following the Transformers upstream and hold off on treating the current Apertus layout as final for now. Once things settle there, we can sync up and see if anything needs adjusting on the llama.cpp side.
